### PR TITLE
Fix macOS M-series (Apple Silicon) support for inference

### PR DESCRIPTION
**Pull request by Infatoshi**
## Summary

Trying to clone and run inference with the pre-trained neural net for the first time on MacBook M4 Max. Found and fixed several issues preventing MPS (Metal Performance Shaders) acceleration.

### Changes

1. **MPS Spectral Norm Support** (BTR.py):
   - PyTorch spectral norm uses torch.vdot which is not implemented for MPS
   - Added monkey-patch that replaces vdot with equivalent (a * b).sum()
   - Benchmarked: **2.2x speedup** for spectral-normed conv layers on MPS vs CPU
   - Numerical accuracy: max difference ~6e-7 (excellent)

2. **Dolphin Launch Fix** (DolphinEnv.py):
   - Changed from open command to direct binary execution on macOS
   - open on macOS does not pass environment variables to child processes
   - Added PYTHONHOME detection for common Python 3.13 locations (uv, Homebrew, python.org)

3. **torchsummary MPS Fix** (BTR_test.py):
   - torchsummary only supports cuda/cpu device strings
   - Skip model summary on MPS with informative message

4. **Setup Script** (scripts/setup_macos_dolphin.py):
   - The bundled Dolphin has Python 3.13.5 embedded but is missing the standard library
   - New script downloads Python 3.13.5 from python.org and copies stdlib into Dolphin bundles
   - Run once after clone_dolphins.py before first use

### Setup Instructions for M-series Macs

After following the existing README setup, run:

    python scripts/setup_macos_dolphin.py
    python BTR_test.py --model_path BTR_MarioKart2000M_82M.model --device mps

### Test Results

Tested on MacBook M4 Max:
- Device: mps
- FPS: 36 (real-time)
- Spectral norm: Working with patch
- Model: Successfully loaded and ran inference

## Test Plan

- [x] Verify spectral norm numerical accuracy vs CPU (max diff < 1e-4)
- [x] Benchmark spectral norm layer performance (MPS 2.2x faster)
- [x] Run full inference with pre-trained model
- [x] Confirm Dolphin scripting engine initializes without errors

### CODE REVIEW

File: scripts/setup_macos_dolphin.py
Comment (unexploredtest) [inline scripts/setup_macos_dolphin.py:1]: nice, I missed that, tho how about including it directly into the bundle by default?

File: scripts/setup_macos_dolphin.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env python3
+"""
+Setup script for macOS M-series (Apple Silicon) Dolphin compatibility.
+
+The bundled Dolphin has Python 3.13.5 embedded but is missing the standard library.
+This script downloads Python 3.13.5 from python.org and copies the stdlib into each
+Dolphin instance's framework bundle.
+
+Run this after clone_dolphins.py and before running BTR.py or BTR_test.py on macOS.
```
Comment (unexploredtest) [inline scripts/setup_macos_dolphin.py]: If we decide to keep the script tho, I think we can call this before `clone_dolphins.py` and then clone dolphin and its bundles.

File: scripts/setup_macos_dolphin.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            check=True
+        )
+
+        # Find the lib directory
+        src_lib = payload_dir / "Versions" / "3.13" / "lib"
+        if not src_lib.exists():
+            print(f"Error: Could not find lib directory at {src_lib}")
+            return
+
+        # Copy to each Dolphin instance
```
Comment (unexploredtest) [inline scripts/setup_macos_dolphin.py]: Copying to each instance shouldn't be necessary if we call before `clone_dolphin.py`

File: BTR.py [hunk 3]
```diff
@@ -22,6 +22,36 @@
...
+                for _ in range(self.n_power_iterations):
+                    self._v = F.normalize(torch.mv(weight_mat.t(), self._u), dim=0, eps=self.eps)
+                    self._u = F.normalize(torch.mv(weight_mat, self._v), dim=0, eps=self.eps)
+        sigma = _metal_vdot(self._u, torch.mv(weight_mat, self._v))
+        return weight / sigma
+
+    parametrizations._SpectralNorm.forward = patched_forward
+
+# Apply patch if MPS is available
+if torch.backends.mps.is_available():
```
Comment (unexploredtest) [inline BTR.py]: Great! Didn't know you could patch it to support vdot, that's why I added an option to disable spectral all together for mps. I think that we should only enable it when the device is mps tho not when it supports mps (so that we wouldn't patch it if we were on CPU with Mac for example), I guess calling it after the creation of device if applicable should do

File: DolphinEnv.py [hunk 4]
```diff
@@ -216,7 +214,20 @@ def create_dolphin(self, i):
...
+            # Find Python 3.13 home for MPS compatibility
+            python_homes = [
+                os.path.expanduser('~/.local/share/uv/python/cpython-3.13.5-macos-aarch64-none'),
+                '/opt/homebrew/Cellar/python@3.13/3.13.5/Frameworks/Python.framework/Versions/3.13',
+                '/opt/homebrew/Cellar/python@3.13/3.13.11/Frameworks/Python.framework/Versions/3.13',
+                '/Library/Frameworks/Python.framework/Versions/3.13',
+            ]
+            for path in python_homes:
+                if os.path.exists(path):
+                    env['PYTHONHOME'] = path
```
Comment (unexploredtest) [inline DolphinEnv.py]: I'm confused, why do we have to change `PYTHONHOME`, didn't we embed python libs into the bundle? Is it not just going to use it? 
I don't have a mac right now unfortunately so can't test it.

**Review by unexploredtest — COMMENTED** [review]
Hey, not the owner but the one who added Mac OS support. Good stuff!

File: DolphinEnv.py (hunk 4 above)
Comment (Infatoshi) [reply DolphinEnv.py]: claude added that for my configuration. oops. disregard

File: BTR.py (hunk 3 above)
Comment (Infatoshi) [reply BTR.py]: perhaps having --device=mps as an arg would be better. im running on M series and assumed others would too

Comment (unexploredtest) [reply BTR.py]: Yeah the arg '--device' does exist and can be set to 'cpu', 'cuda', 'mps' etc, you can check if it's mps right after this line and call the function:
https://github.com/VIPTankz/Wii-RL/blob/5053b8b8980ed4ef047b60fd7180b443c00c9aff/BTR.py#L1251

**Comment by Infatoshi** [comment]
## Addressing PR feedback

Thanks for the review @unexploredtest! I've pushed commits addressing your points:

### Setup script order (setup_macos_dolphin.py)
Updated the script to only target `dolphin0` and run **before** `clone_dolphins.py`. Now the stdlib gets copied once to dolphin0, then automatically included when cloning to other instances. Much cleaner than copying to each one separately.

Including it directly in the bundle would be ideal long-term, but this approach works without requiring changes to the Dolphin build process.

### MPS spectral norm patch (BTR.py)
Moved the vdot patch to only apply when `--device mps` is actually used, not just when MPS is available on the system. This keeps CPU mode untouched on Macs.

### PYTHONHOME (DolphinEnv.py)
Removed the PYTHONHOME block entirely - it was added for my local setup but shouldn't be needed if the stdlib is properly embedded via the setup script.
